crypto/ed25519.go

```diff
@@ -36,40 +41,90 @@ func GenerateEd25519Keypair() (*Ed25519Keypair, error) {
 
 // NewEd25519PublicKey returns an ed25519 public key that consists of the input bytes
 // Input length must be 32 bytes
-func NewEd25519PublicKey(in []byte) (ed25519.PublicKey, error) {
+func NewEd25519PublicKey(in []byte) (*Ed25519PublicKey, error) {
 	if len(in) != 32 {
 		return nil, fmt.Errorf("cannot create public key: input is not 32 bytes")
 	}
 
-	return ed25519.PublicKey(in), nil
+	pub := Ed25519PublicKey(ed25519.PublicKey(in))
+	return &pub, nil
 }
 
 // NewEd25519PrivateKey returns an ed25519 private key that consists of the input bytes
 // Input length must be 64 bytes
-func NewEd25519PrivateKey(in []byte) (ed25519.PrivateKey, error) {
+func NewEd25519PrivateKey(in []byte) (*Ed25519PrivateKey, error) {
 	if len(in) != 64 {
 		return nil, fmt.Errorf("cannot create private key: input is not 64 bytes")
 	}
 
-	return ed25519.PrivateKey(in), nil
+	priv := Ed25519PrivateKey(ed25519.PrivateKey(in))
+	return &priv, nil
+}
+
+// Verify returns true if the signature is valid for the given message and public key, false otherwise
+func Verify(pub *Ed25519PublicKey, msg, sig []byte) bool {
+	return ed25519.Verify(ed25519.PublicKey(*pub), msg, sig)
 }
 
-// Sign uses the keypair to sign the message using the ed25519 signature algorith
+// Sign uses the keypair to sign the message using the ed25519 signature algorithm
 func (kp *Ed25519Keypair) Sign(msg []byte) []byte {
-	return ed25519.Sign(kp.private, msg)
+	return ed25519.Sign(ed25519.PrivateKey(*kp.private), msg)
 }
 
 // Public returns the keypair's public key
-func (kp *Ed25519Keypair) Public() ed25519.PublicKey {
+func (kp *Ed25519Keypair) Public() PublicKey {
 	return kp.public
 }
 
 // Private returns the keypair's private key
-func (kp *Ed25519Keypair) Private() ed25519.PrivateKey {
+func (kp *Ed25519Keypair) Private() PrivateKey {
 	return kp.private
 }
 
-// Verify returns true if the signature is valid for the given message and public key, false otherwise
-func Verify(pub ed25519.PublicKey, msg, sig []byte) bool {
-	return ed25519.Verify(pub, msg, sig)
+// Sign uses the ed25519 signature algorithm to sign the message
+func (k *Ed25519PrivateKey) Sign(msg []byte) []byte {
+	return ed25519.Sign(ed25519.PrivateKey(*k), msg)
+}
+
+// Public returns the public key corresponding to the ed25519 private key
+func (k *Ed25519PrivateKey) Public() PublicKey {
+	kp := NewEd25519Keypair(ed25519.PrivateKey(*k))
+	return kp.Public()
+}
+
+// Encode returns the bytes underlying the Ed25519PrivateKey
+func (k *Ed25519PrivateKey) Encode() []byte {
+	return []byte(ed25519.PrivateKey(*k))
+}
+
+// Decode turns the input bytes into a Ed25519PrivateKey
+// the input must be 32 bytes, or the function will return an error
```

Comment:
32 bytes or 64 bytes? 

`NewEd25519PrivateKey` checks for 64 bytes...

Reply:
you are right, I've updated the comment